### PR TITLE
AEA-1037: Refactor Spine response translation.

### PR DESCRIPTION
**Pull request by mfjarvis**
Refactor Spine response translation to share code and make it easier to add new responses.
Add some fixes to handle optional fields in release response messages.
Add reversibility tests (skipped because they won't pass in their current state, but they are useful to run manually).
Work around a bug in our XML library where XML characters are not escaped in attributes.
Create a second tsconfig.json file for builds, and change the main file to include tests (fixes IDE warnings)

### CODE REVIEW

File: coordinator/src/services/translation/response/spine-response-handler.ts [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+    sendMessagePayload: hl7V3.SendMessagePayload<T>,
+    logger: pino.Logger
+  ): TranslatedSpineResponse {
+    const errorCodes = this.extractErrorCodes(sendMessagePayload)
+    return SpineResponseHandler.handleErrorOrRejectionResponse(errorCodes, logger)
+  }
+
+  protected handleSuccessResponse(
+    sendMessagePayload: hl7V3.SendMessagePayload<T>,
+    logger: pino.Logger
```
Comment (dan-jones-iw) [inline coordinator/src/services/translation/response/spine-response-handler.ts:133]: doesnt seem like the logger is used here or in any of the extension classes

File: coordinator/tests/services/translation/reversibility.spec.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+import {toArray} from "../../../src/services/translation/common"
+
+describe("translations are reversible", () => {
+  const bundle = TestResources.specification[0].fhirMessageSigned
+  const parentPrescription = convertParentPrescription(bundle)
+  ensureLineItemStatus(parentPrescription)
+  const originalBundle = JSON.parse(LosslessJson.stringify(bundle))
+  const translatedBundle = createInnerBundle(parentPrescription, uuid.v4())
+
+  test.skip.each([
```
Comment (dan-jones-iw) [inline coordinator/tests/services/translation/reversibility.spec.ts:18]: any updates on the skip? are we keeping this in for later?

File: coordinator/tests/services/translation/xml.spec.ts [hunk 3]
```diff
@@ -1,8 +1,82 @@
...
+import {writeXmlStringCanonicalized, writeXmlStringPretty} from "../../../src/services/serialisation/xml"
 import * as TestResources from "../../resources/test-resources"
 
 test("writeXmlStringCanonicalized returns correct value", () => {
   const actualOutput = writeXmlStringCanonicalized(TestResources.examplePrescription1.hl7V3SignatureFragments)
   const expectedOutput = TestResources.examplePrescription1.hl7V3FragmentsCanonicalized
   expect(actualOutput).toEqual(expectedOutput)
 })
+
+//TODO - add more tests to prove that XML is correctly canonicalized
```
Comment (dan-jones-iw) [inline coordinator/tests/services/translation/xml.spec.ts:10]: assume this is done?

File: coordinator/tsconfig-build.json [hunk 4]
```diff
@@ -0,0 +1,6 @@
+{
+  "extends": "./tsconfig.json",
+  "exclude": [
```
Comment (dan-jones-iw) [inline coordinator/tsconfig-build.json:3]: maybe better to use include vs exclude in build config? if we ever add any ts files outside of `/src` and `/tests` will be automatically built (not rly the biggest deal)

File: coordinator/tests/services/translation/xml.spec.ts (hunk 3 above)
Comment (dan-jones-iw) [reply coordinator/tests/services/translation/xml.spec.ts:10]: do we have tests around new `readXml` stuff?

**Review by dan-jones-iw — APPROVED** [review]

File: coordinator/src/services/translation/response/spine-response-handler.ts [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+    return SpineResponseHandler.createSuccessResponse()
+  }
+
+  /* eslint-enable @typescript-eslint/no-unused-vars */
+}
+
+export class CancelResponseHandler extends SpineResponseHandler<hl7V3.CancellationResponseRoot> {
+  translator: (cancelResponse: hl7V3.CancellationResponse) => fhir.Bundle
+
+  constructor(interactionId: string, translator: (cancelResponse: hl7V3.CancellationResponse) => fhir.Bundle) {
```
[the remark this answers is not in the document]
Comment (mfjarvis) [reply coordinator/src/services/translation/response/spine-response-handler.ts:144]: Done for ease of testing. Means we can pass in a mock and our tests aren't dependent on the entire cancel translation code

File: coordinator/src/services/translation/response/spine-response-handler.ts (hunk 1 above)
Comment (mfjarvis) [reply coordinator/src/services/translation/response/spine-response-handler.ts:133]: Future proofing (/ over engineering) for when we want to log stuff inside our translation code

File: coordinator/tests/services/translation/reversibility.spec.ts (hunk 2 above)
Comment (mfjarvis) [reply coordinator/tests/services/translation/reversibility.spec.ts:18]: Yeah I'm going to address the test failures in a new PR :+1:

File: coordinator/tests/services/translation/xml.spec.ts (hunk 3 above)
Comment (mfjarvis) [reply coordinator/tests/services/translation/xml.spec.ts:10]: No, we've still only got the one test for canonicalization and it's not a very good test. Spotted this but didn't want to fix as part of this PR

Comment (mfjarvis) [reply coordinator/tests/services/translation/xml.spec.ts:10]: Will add the tests for the strip namespace stuff though, good spot

File: coordinator/tsconfig-build.json (hunk 4 above)
Comment (mfjarvis) [reply coordinator/tsconfig-build.json:3]: Depends on whether `include` overrides or adds to the `include` in the file it's extending. I'll look into it

Comment (mfjarvis) [reply coordinator/tsconfig-build.json:3]: It does override. Changed.

File: coordinator/tests/services/translation/xml.spec.ts (hunk 3 above)
Comment (mfjarvis) [reply coordinator/tests/services/translation/xml.spec.ts:10]: Added tests.

File: coordinator/src/services/translation/response/spine-response-handler.ts (hunk 1 above)
Comment (dan-jones-iw) [reply coordinator/src/services/translation/response/spine-response-handler.ts:133]: lol okay :+1:

File: coordinator/src/services/translation/response/spine-response-handler.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (dan-jones-iw) [reply coordinator/src/services/translation/response/spine-response-handler.ts:144]: makes sense

**Review by dan-jones-iw — APPROVED** [review]

**Pull request merged**